### PR TITLE
[flink] Describe Function autocomplete for flink

### PR DESCRIPTION
**Pull request by Akhilsnaik**
**What changes were proposed in this pull request?**

https://github.com/cloudera/hue/issues/1683 : [flink] Describe Function autocomplete for flink

**How was this patch tested?**

- Tested manually in Flink Editor UI , the describer autocomplete is working as needed. attached screenshots

Screenshot 2021-01-24 at 10 15 39 PM

Screenshot 2021-01-24 at 10 27 38 PM

### CODE REVIEW

**Comment by romainr** [comment]
Really nice!!

Main comment is shouldn't we move the files into https://github.com/cloudera/hue/tree/master/desktop/core/src/desktop/js/parse/jison/sql/calcite instead of flink?

I already see that we could reuse those autocomplete pieces in Phoenix and more.

**Comment by Akhilsnaik** [comment]
> Really nice!!
> 
> Main comment is shouldn't we move the files into https://github.com/cloudera/hue/tree/master/desktop/core/src/desktop/js/parse/jison/sql/calcite instead of flink?
> 
> I already see that we could reuse those autocomplete pieces in Phoenix and more.

the Syntax of Calcite and Flink seems to be different for DESCRIBE
Calcite
Screenshot 2021-01-25 at 8 33 06 AM

Flink : https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/sql/describe.html#syntax

Flink can only be subset of Calcite  ,

**Comment by romainr** [comment]
Is this PR about DESCRIBE functions (cf. title) or DB/table? Maybe adding a test would demo it and is good practice BTW.

**Comment by Akhilsnaik** [comment]
> Really nice!!
> 
> Main comment is shouldn't we move the files into https://github.com/cloudera/hue/tree/master/desktop/core/src/desktop/js/parse/jison/sql/calcite instead of flink?
> 
> I already see that we could reuse those autocomplete pieces in Phoenix and more.

I have moved the describe_table.json into calcite and reused in flink.

File: desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ ;
+
+DataDefinition_EDIT
+ : DescribeStatement_EDIT
+ ;
+
+DescribeStatement
+ : 'DESCRIBE' SchemaQualifiedTableIdentifier
+   {
+     parser.addTablePrimary($3);
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison:28]: $3 should be $2, first token is $1.

File: desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+DescribeStatement_EDIT
+ : 'DESCRIBE' 'CURSOR'
+   {
+     parser.suggestTables();
+     parser.suggestDatabases({ appendDot: true });
+   }
+ | 'DESCRIBE' SchemaQualifiedTableIdentifier_EDIT
+ | 'DESCRIBE' 'CURSOR' SchemaQualifiedTableIdentifier
+   {
+     parser.addTablePrimary($4);
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison:41]: Same, $4 -> $3

**Review by JohanAhlen — COMMENTED** [review]
Nice one!

Note that Flink also accepts an optional catalog name:

  `DESCRIBE [catalog_name.][db_name.]table_name`

For the syntax parser to accept it you can add an extra line in the DescribeStatement rule

 ```
: 'DESCRIBE' SchemaQualifiedTableIdentifier
  ...
| 'DESCRIBE' RegularOrBacktickedIdentifier '.' SchemaQualifiedTableIdentifier
  {
    parser.addTablePrimary($4);
  }
```

For autocomplete I'm not sure if we support catalog in the autocomplete api endpoints, fyi @romainr, so can probably wait with that.

As Romain mentioned, add some tests, take a look at prestoAutocompleteParser.Describe.test.js for reference.

File: desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison (hunk 1 above)
Comment (Akhilsnaik) [reply desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison:32]: this has been corrected

File: jest.config.js [hunk 3]
```diff
@@ -21,8 +21,6 @@ module.exports = {
     '<rootDir>/desktop/core/src/desktop/js/parse/sql/phoenix/test',
     '<rootDir>/desktop/core/src/desktop/js/parse/sql/druid/test',
     '<rootDir>/desktop/core/src/desktop/js/parse/sql/presto/test',
-    '<rootDir>/desktop/core/src/desktop/js/parse/sql/calcite/test',
```
Comment (Akhilsnaik) [inline jest.config.js:24]: @JohanAhlen  I have removed the jest.config for calcite and Flink for it to run

File: desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ ;
+
+DescribeStatement
+ : 'DESCRIBE' SchemaQualifiedTableIdentifier
+   {
+     parser.addTablePrimary($2);
+   }
+ | 'DESCRIBE' RegularOrBacktickedIdentifier '.' SchemaQualifiedTableIdentifier
+   {
+     parser.addTablePrimary($3);
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison:32]: parser.addTablePrimary($4);

index i 1-based

'DESCRIBE' -> $1
RegularOrBacktickedIdentifier -> $2
'.' -> $3
SchemaQualifiedIdentifier -> $4

File: desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+     parser.suggestDatabases({ appendDot: true });
+   }
+ | 'DESCRIBE' SchemaQualifiedTableIdentifier_EDIT
+ | 'DESCRIBE' 'CURSOR' SchemaQualifiedTableIdentifier
+   {
+     parser.addTablePrimary($3);
+   }
+ | 'DESCRIBE' RegularOrBacktickedIdentifier '.' SchemaQualifiedTableIdentifier 
+   {
+     parser.addTablePrimary($5);
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/parse/jison/sql/calcite/describe/describe_table.jison:49]: $4

File: desktop/core/src/desktop/js/parse/sql/calcite/test/calciteutocompleteParser.Describe.test.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import calciteAutocompleteParser from '../calciteAutocompleteParser';
+
+describe('calciteAutocompleteParser.js DESCRIBE statements', () => {
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/js/parse/sql/calcite/test/calciteutocompleteParser.Describe.test.js:19]: Nice!

File: jest.config.js (hunk 3 above)
Comment (JohanAhlen) [reply jest.config.js:24]: It's there to reduce the CI build time. For now keep it as ignored in this list, we'll need to to split tests per jison file instead to get rid of the duplication.

**Review by JohanAhlen — APPROVED** [review]

**Review by JohanAhlen — APPROVED** [review]

**Comment by romainr** [comment]
Nice!

Will play with it in today's gethue.demo refresh!

**Pull request merged**